docs/advanced/smart_ptrs.rst

```diff
@@ -1,126 +1,120 @@
-Smart pointers
-##############
+Smart pointers & ``py::class_``
+###############################
 
-std::unique_ptr
-===============
+The binding generator for classes, ``py::class_``, can be passed a template
+type that denotes a special *holder* type that is used to manage references to
+the object. If no such holder type template argument is given, the default for
+a type ``T`` is ``std::unique_ptr<T>``.
 
-Given a class ``Example`` with Python bindings, it's possible to return
-instances wrapped in C++11 unique pointers, like so
+.. note::
 
-.. code-block:: cpp
+    A ``py::class_`` for a given C++ type ``T`` — and all its derived types —
+    can only use a single holder type.
 
-    std::unique_ptr<Example> create_example() { return std::unique_ptr<Example>(new Example()); }
 
-.. code-block:: cpp
+.. _smart_holder:
 
-    m.def("create_example", &create_example);
+``py::smart_holder``
+====================
 
-In other words, there is nothing special that needs to be done. While returning
-unique pointers in this way is allowed, it is *illegal* to use them as function
-arguments. For instance, the following function signature cannot be processed
-by pybind11.
+Starting with pybind11v3, ``py::smart_holder`` is built into pybind11. It is
+the recommended ``py::class_`` holder for most situations. However, for
```

Comment:
Did you mean `py::classh` here?

Reply:
No this is correct here, although I can see that it is difficult to read in the github diff view. Sorry I forgot to point out in the PR description how to get to the updated doc preview as it will look on readthedocs after the PR is merged:

https://pybind11--5542.org.readthedocs.build/en/5542/

I'll add that link in the PR description now.

Reply:
Sorry, newbee here. I am confused by the "Starting with pybind11v3" part. The latest release appears to be 2.13.6 from late 2024 and this PR was merged 3 weeks ago, but the online docs seem to imply `smart_holder` should currently be used. If this is the case, do I need to access it via `master`? (currently I'm consuming pybind11 via vcpkg)

Reply:
There should be several docs. The ones you normally see (https://pybind11.readthedocs.io/en/stable/) should be the docs from 2.13.6, and should not mention `smart_holder`. If you go to https://pybind11.readthedocs.io/en/latest, then you'll see a warning about the docs being for a development version, and those does do have `smart_holder` in them (and a v3 as well).

Reply:
> Sorry, newbee here. I am confused by the "Starting with pybind11v3" part. The latest release appears to be 2.13.6 from late 2024 and this PR was merged 3 weeks ago, but the online docs seem to imply `smart_holder` should currently be used. If this is the case, do I need to access it via `master`? (currently I'm consuming pybind11 via vcpkg which is tied to the releases)

The docs for releases are here:

https://pybind11.readthedocs.io/en/stable/

You seem to be looking at the docs for master branch. (?)

Unless you need unreleased features or fixes, there is no need to switch to using the master branch.

Reply:
Thank you! I am looking at [](https://pybind11.readthedocs.io/en/latest/index.html), I'll switch to stable docs. I think I must have ended up on `latest` docs by clicking the doc link in the github readme which is master branch.